Implement some of MSVC's intrinsics, for ImportC.

just-harry opened the pull request:
This PR implements all but three of the MSVC intrinsics listed here: https://web.archive.org/web/20240412171516/https://learn.microsoft.com/en-ie/cpp/intrinsics/alphabetical-listing-of-intrinsic-functions?view=msvc-170, and a handful of undocumented intrinsics.

The three unimplemented intrinsics are `_AddressOfReturnAddress`, `__getcallerseflags`, and `_ReturnAddress`, because they need compiler support for their implementation.

<details>
<summary>The implemented intrinsics are listed in this expando.</summary>

- `__assume`
- `__umulh`
- `__mulh`
- `_umul128`
- `_mul128`
- `__emul`
- `__emulu`
- `_div128`
- `_udiv128`
- `_div64`
- `_udiv64`
- `__cpuid`
- `__cpuidex`
- `_cvt_ftoi_fast`
- `_cvt_ftoll_fast`
- `_cvt_ftoui_fast`
- `_cvt_ftoull_fast`
- `_cvt_dtoi_fast`
- `_cvt_dtoll_fast`
- `_cvt_dtoui_fast`
- `_cvt_dtoull_fast`
- `_cvt_ftoi_sat`
- `_cvt_ftoll_sat`
- `_cvt_ftoui_sat`
- `_cvt_ftoull_sat`
- `_cvt_dtoi_sat`
- `_cvt_dtoll_sat`
- `_cvt_dtoui_sat`
- `_cvt_dtoull_sat`
- `_cvt_ftoi_sent`
- `_cvt_ftoll_sent`
- `_cvt_ftoui_sent`
- `_cvt_ftoull_sent`
- `_cvt_dtoi_sent`
- `_cvt_dtoui_sent`
- `_cvt_dtoll_sent`
- `_cvt_dtoull_sent`
- `__readgsbyte`
- `__readgsword`
- `__readgsdword`
- `__readgsqword`
- `__writegsbyte`
- `__writegsword`
- `__writegsdword`
- `__writegsqword`
- `__addgsbyte`
- `__addgsword`
- `__addgsdword`
- `__addgsqword`
- `__incgsbyte`
- `__incgsword`
- `__incgsdword`
- `__incgsqword`
- `__readfsbyte`
- `__readfsword`
- `__readfsdword`
- `__readfsqword`
- `__writefsbyte`
- `__writefsword`
- `__writefsdword`
- `__writefsqword`
- `__addfsbyte`
- `__addfsword`
- `__addfsdword`
- `__incfsbyte`
- `__incfsword`
- `__incfsdword`
- `__debugbreak`
- `__fastfail`
- `__faststorefence`
- `_disable`
- `_enable`
- `_interlockedadd [Undocumented]`
- `_interlockedadd64 [Undocumented]`
- `_InterlockedAdd`
- `_InterlockedAdd_acq`
- `_InterlockedAdd_rel`
- `_InterlockedAdd_nf`
- `_InterlockedAdd64`
- `_InterlockedAdd64_acq`
- `_InterlockedAdd64_rel`
- `_InterlockedAdd64_nf`
- `_InterlockedAddLargeStatistic`
- `_InterlockedAnd`
- `_InterlockedAnd8`
- `_InterlockedAnd16`
- `_interlockedand64 [Undocumented]`
- `_InterlockedAnd_acq`
- `_InterlockedAnd_rel`
- `_InterlockedAnd_nf`
- `_InterlockedAnd8_acq`
- `_InterlockedAnd8_rel`
- `_InterlockedAnd8_nf`
- `_InterlockedAnd16_acq`
- `_InterlockedAnd16_rel`
- `_InterlockedAnd16_nf`
- `_InterlockedAnd64_acq`
- `_InterlockedAnd64_rel`
- `_InterlockedAnd64_nf`
- `_InterlockedAnd64`
- `_InterlockedAnd_np`
- `_InterlockedAnd8_np`
- `_InterlockedAnd16_np`
- `_InterlockedAnd64_np`
- `_InterlockedAnd64_HLEAcquire`
- `_InterlockedAnd64_HLERelease`
- `_InterlockedAnd_HLEAcquire`
- `_InterlockedAnd_HLERelease`
- `_interlockedbittestandreset`
- `_interlockedbittestandreset64`
- `_interlockedbittestandreset_HLEAcquire`
- `_interlockedbittestandreset_HLERelease`
- `_interlockedbittestandreset64_HLEAcquire`
- `_interlockedbittestandreset64_HLERelease`
- `_interlockedbittestandreset_acq`
- `_interlockedbittestandreset_rel`
- `_interlockedbittestandreset_nf`
- `_interlockedbittestandreset64_acq`
- `_interlockedbittestandreset64_rel`
- `_interlockedbittestandreset64_nf`
- `_interlockedbittestandset`
- `_interlockedbittestandset64`
- `_interlockedbittestandset_HLEAcquire`
- `_interlockedbittestandset_HLERelease`
- `_interlockedbittestandset64_HLEAcquire`
- `_interlockedbittestandset64_HLERelease`
- `_interlockedbittestandset_acq`
- `_interlockedbittestandset_rel`
- `_interlockedbittestandset_nf`
- `_interlockedbittestandset64_acq`
- `_interlockedbittestandset64_rel`
- `_interlockedbittestandset64_nf`
- `_InterlockedCompareExchange`
- `_InterlockedCompareExchange8`
- `_InterlockedCompareExchange16`
- `_InterlockedCompareExchange64`
- `_InterlockedCompareExchange_HLEAcquire`
- `_InterlockedCompareExchange_HLERelease`
- `_InterlockedCompareExchange64_HLEAcquire`
- `_InterlockedCompareExchange64_HLERelease`
- `_InterlockedCompareExchange_np`
- `_InterlockedCompareExchange16_np`
- `_InterlockedCompareExchange64_np`
- `_InterlockedCompareExchange_acq`
- `_InterlockedCompareExchange_rel`
- `_InterlockedCompareExchange_nf`
- `_InterlockedCompareExchange8_acq`
- `_InterlockedCompareExchange8_rel`
- `_InterlockedCompareExchange8_nf`
- `_InterlockedCompareExchange16_acq`
- `_InterlockedCompareExchange16_rel`
- `_InterlockedCompareExchange16_nf`
- `_InterlockedCompareExchange64_acq`
- `_InterlockedCompareExchange64_rel`
- `_InterlockedCompareExchange64_nf`
- `_InterlockedCompareExchange128`
- `_InterlockedCompareExchange128_np`
- `_InterlockedCompareExchange128_acq`
- `_InterlockedCompareExchange128_rel`
- `_InterlockedCompareExchange128_nf`
- `_InterlockedCompareExchangePointer`
- `_InterlockedCompareExchangePointer_HLEAcquire`
- `_InterlockedCompareExchangePointer_HLERelease`
- `_InterlockedCompareExchangePointer_np`
- `_InterlockedCompareExchangePointer_acq`
- `_InterlockedCompareExchangePointer_rel`
- `_InterlockedCompareExchangePointer_nf`
- `_InterlockedDecrement`
- `_InterlockedDecrement16`
- `_interlockeddecrement64`
- `_InterlockedDecrement64`
- `_InterlockedDecrement_acq`
- `_InterlockedDecrement_rel`
- `_InterlockedDecrement_nf`
- `_InterlockedDecrement16_acq`
- `_InterlockedDecrement16_rel`
- `_InterlockedDecrement16_nf`
- `_InterlockedDecrement64_acq`
- `_InterlockedDecrement64_rel`
- `_InterlockedDecrement64_nf`
- `_InterlockedExchange`
- `_InterlockedExchange8`
- `_InterlockedExchange16`
- `_interlockedexchange64`
- `_InterlockedExchange64`
- `_InterlockedExchange_HLEAcquire`
- `_InterlockedExchange_HLERelease`
- `_InterlockedExchange64_HLEAcquire`
- `_InterlockedExchange64_HLERelease`
- `_InterlockedExchange_acq`
- `_InterlockedExchange_rel`
- `_InterlockedExchange_nf`
- `_InterlockedExchange8_acq`
- `_InterlockedExchange8_rel`
- `_InterlockedExchange8_nf`
- `_InterlockedExchange16_acq`
- `_InterlockedExchange16_rel`
- `_InterlockedExchange16_nf`
- `_InterlockedExchange64_acq`
- `_InterlockedExchange64_rel`
- `_InterlockedExchange64_nf`
- `_InterlockedExchangeAdd`
- `_InterlockedExchangeAdd8`
- `_InterlockedExchangeAdd16`
- `_interlockedexchangeadd64 [Undocumented]`
- `_InterlockedExchangeAdd64`
- `_InterlockedExchangeAdd_HLEAcquire`
- `_InterlockedExchangeAdd_HLERelease`
- `_InterlockedExchangeAdd64_HLEAcquire`
- `_InterlockedExchangeAdd64_HLERelease`
- `_InterlockedExchangeAdd_acq`
- `_InterlockedExchangeAdd_rel`
- `_InterlockedExchangeAdd_nf`
- `_InterlockedExchangeAdd8_acq`
- `_InterlockedExchangeAdd8_rel`
- `_InterlockedExchangeAdd8_nf`
- `_InterlockedExchangeAdd16_acq`
- `_InterlockedExchangeAdd16_rel`
- `_InterlockedExchangeAdd16_nf`
- `_InterlockedExchangeAdd64_acq`
- `_InterlockedExchangeAdd64_rel`
- `_InterlockedExchangeAdd64_nf`
- `_InterlockedExchangePointer`
- `_InterlockedExchangePointer_HLEAcquire`
- `_InterlockedExchangePointer_HLERelease`
- `_InterlockedExchangePointer_acq`
- `_InterlockedExchangePointer_rel`
- `_InterlockedExchangePointer_nf`
- `_InterlockedIncrement`
- `_InterlockedIncrement16`
- `_interlockedincrement64 [Undocumented]`
- `_InterlockedIncrement64`
- `_InterlockedIncrement_acq`
- `_InterlockedIncrement_rel`
- `_InterlockedIncrement_nf`
- `_InterlockedIncrement16_acq`
- `_InterlockedIncrement16_rel`
- `_InterlockedIncrement16_nf`
- `_InterlockedIncrement64_acq`
- `_InterlockedIncrement64_rel`
- `_InterlockedIncrement64_nf`
- `_InterlockedOr`
- `_InterlockedOr8`
- `_InterlockedOr16`
- `_interlockedor64 [Undocumented]`
- `_InterlockedOr_acq`
- `_InterlockedOr_rel`
- `_InterlockedOr_nf`
- `_InterlockedOr8_acq`
- `_InterlockedOr8_rel`
- `_InterlockedOr8_nf`
- `_InterlockedOr16_acq`
- `_InterlockedOr16_rel`
- `_InterlockedOr16_nf`
- `_InterlockedOr64_acq`
- `_InterlockedOr64_rel`
- `_InterlockedOr64_nf`
- `_InterlockedOr64`
- `_InterlockedOr_np`
- `_InterlockedOr8_np`
- `_InterlockedOr16_np`
- `_InterlockedOr64_np`
- `_InterlockedOr64_HLEAcquire`
- `_InterlockedOr64_HLERelease`
- `_InterlockedOr_HLEAcquire`
- `_InterlockedOr_HLERelease`
- `_InterlockedXor`
- `_InterlockedXor8`
- `_InterlockedXor16`
- `_interlockedxor64 [Undocumented]`
- `_InterlockedXor_acq`
- `_InterlockedXor_rel`
- `_InterlockedXor_nf`
- `_InterlockedXor8_acq`
- `_InterlockedXor8_rel`
- `_InterlockedXor8_nf`
- `_InterlockedXor16_acq`
- `_InterlockedXor16_rel`
- `_InterlockedXor16_nf`
- `_InterlockedXor64_acq`
- `_InterlockedXor64_rel`
- `_InterlockedXor64_nf`
- `_InterlockedXor64`
- `_InterlockedXor_np`
- `_InterlockedXor8_np`
- `_InterlockedXor16_np`
- `_InterlockedXor64_np`
- `_InterlockedXor64_HLEAcquire`
- `_InterlockedXor64_HLERelease`
- `_InterlockedXor_HLEAcquire`
- `_InterlockedXor_HLERelease`
- `__inbyte`
- `__inword`
- `__indword`
- `__outbyte`
- `__outword`
- `__outdword`
- `__inbytestring`
- `__inwordstring`
- `__indwordstring`
- `__outbytestring`
- `__outwordstring`
- `__outdwordstring`
- `__int2c`
- `__invlpg`
- `__lidt`
- `__ll_lshift`
- `__ll_rshift`
- `__ull_rshift`
- `__lzcnt16`
- `__lzcnt`
- `__lzcnt64`
- `_mm_cvtsi64x_ss`
- `_mm_cvtss_si64x`
- `_mm_cvttss_si64x`
- `_mm_extract_si64`
- `_mm_extracti_si64`
- `_mm_insert_si64`
- `_mm_inserti_si64`
- `_mm_stream_sd`
- `_mm_stream_ss`
- `_mm_stream_si64x`
- `__movsb`
- `__movsw`
- `__movsd`
- `__movsq`
- `__noop`
- `__nop`
- `__popcnt16`
- `__popcnt`
- `__popcnt64`
- `__rdtsc`
- `__rdtscp`
- `__readcr0`
- `__readcr2`
- `__readcr3`
- `__readcr4`
- `__readcr8`
- `__readdr`
- `__readeflags`
- `__readmsr`
- `__readpmc`
- `__segmentlimit`
- `__shiftleft128`
- `__shiftright128`
- `__sidt`
- `__stosb`
- `__stosw`
- `__stosd`
- `__stosq`
- `__svm_clgi`
- `__svm_invlpga`
- `__svm_skinit`
- `__svm_stgi`
- `__svm_vmload`
- `__svm_vmrun`
- `__svm_vmsave`
- `__ud2`
- `__vmx_off`
- `__vmx_on`
- `__vmx_vmclear`
- `__vmx_vmlaunch`
- `__vmx_vmptrld`
- `__vmx_vmptrst`
- `__vmx_vmread`
- `__vmx_vmresume`
- `__vmx_vmwrite`
- `__wbinvd`
- `__writecr0`
- `__writecr2`
- `__writecr3`
- `__writecr4`
- `__writecr8`
- `__writedr`
- `__writeeflags`
- `__writemsr`
- `_ReadBarrier`
- `_WriteBarrier`
- `_ReadWriteBarrier`
- `_BitScanForward`
- `_BitScanReverse`
- `_BitScanForward64`
- `_BitScanReverse64`
- `_bittest`
- `_bittestandcomplement`
- `_bittestandreset`
- `_bittestandset`
- `_bittest64`
- `_bittestandcomplement64`
- `_bittestandreset64`
- `_bittestandset64`
- `_byteswap_uint64`
- `_byteswap_ulong`
- `_byteswap_ushort`
- `_lrotr`
- `_lrotl`
- `_rotr`
- `_rotl`
- `_rotr64`
- `_rotl64`
- `_rotr16`
- `_rotl16`
- `_rotr8`
- `_rotl8`
</details>

---

These implementations aim to be as compatible with the MSVC intrinsics as is possible—adhering to Hyrum's Law.

Separate implementations are provided for DMD, LDC, and GDC, and for x86, x86-64, AArch64, and ARM.
Almost all the intrinsics are implemented in D, except for `__assume` which is a C macro.

Every intrinsic, where possible, has a CTFE-compatible code-path. \
It all compiles with or without DIP1000 being enabled. \
Care has been taken to ensure that none of the implementations rely on DRuntime, so that these work in BetterC.

---

Regarding the `_cvt_` family of functions: by default MSVC will generate code that uses SSE2 instructions, even for 32-bit targets, which means that for 32-bit targets the `_cvt_` functions will use SSE2.
This is contrary to DMD's usual behaviour of using x87 for 32-bit Windows.
For their reimplementations, I've used SSE2 anyway for 32-bit targets for DMD, as doing otherwise would constitute a change in behaviour, as x87 FP-exceptions are different from SIMD FP-exceptions (and, I think the SSE2 and x87 versions return different results).
The oldest of the `_cvt_` functions was introduced in the May of 2021, so I think it's almost certain that any code using them will be targeting at-least SSE2 anyway.

Additionally, I've written a program that tests that the `_cvt_` implementations return identical results to the MSVC implementations for all float values, and for ~402,653,184 double values (except for `_cvt_ftoi_fast` and `_cvt_ftoi_sent` on 32-bit targets, as they cause an internal compiler error in MSVC); it also tests that `ctfeX86RoundLongToFloat` and `ctfeX86RoundFloatToLong` produce the same results as the hardware. \
It relies on Phobos and MSVC, so I don't really know what to do with it other than link to it here: https://github.com/just-harry/float-fuzzing-for-msvc-intrinsics

---

A few of the intrinsics can be used only in kernel-mode, so unittests have been omitted for them, as I don't think we have any infrastructure in place for testing in kernel-mode.

Some of the intrinsics terminate the program, so their unittests have been wrapped in a `version (none)`, others rely on specific compiler optimisations, so they too have been wrapped in a `version (none)`.

---

I've split the intrinsics up into a few dozen commits to try and alleviate the whole 13,000-lines-of-codes-all-at-once thing. \
(I staged them after-the-fact, so a few braces may be out-of-place in the intermediate commits.)

---

The intrinsics have been placed in their own files, separate from the existing ImportC builtin files, to try and avoid crowding the builtin files.

Currently, the ImportC builtins are imported conditionally, based on some heuristics.
One of those heuristics is if any identifier beginning with two underscores is used – I've changed that one to instead trigger on a single leading underscore as many of MSVC's intrinsics begin with only one underscore.

---

One notable header that can be successfully included by ImportC, with these intrinsics implemented, is `windows.h`.

---

P.S. If this is outside the purview of ImportC: that's fine, I'll just publish this as a library instead.

just-harry committed:
Add files for MSVC intrinsic implementations.

just-harry committed:
Scaffold some versions for the MSVC intrinsic implementations.

just-harry committed:
Implement the MSVC intrinsics for extended-width multiplication.

just-harry committed:
Implement the MSVC intrinsics for extended-width division.

just-harry committed:
Implement the MSVC intrinsics for the x86 `cpuid` instruction.

just-harry committed:
Define `gccBuiltins` privately for the MSVC intrinsic implementations.

just-harry committed:
Implement the MSVC intrinsics for fast float conversion.

just-harry committed:
Implement the MSVC intrinsics for fast double conversion.

just-harry committed:
Implement the MSVC intrinsics for saturating float conversion.

just-harry committed:
Implement the MSVC intrinsics for saturating double conversion.

just-harry committed:
Implement the MSVC intrinsics for sentinel-on-invalid float conversion.

just-harry committed:
Implement the MSVC intrinsics for sentinel-on-invalid double conversion.

just-harry committed:
Implement the `__halt` MSVC intrinsic.

just-harry committed:
Implement the `<op><g/f>s<size>` MSVC intrinsics.

just-harry committed:
Implement the `__debugbreak` MSVC intrinsic.

just-harry committed:
Implement the `__fastfail` MSVC intrinsic.

just-harry committed:
Implement the `__faststorefence` MSVC intrinsic.

just-harry committed:
Implement the `_disable` MSVC intrinsic.

just-harry committed:
Implement the `_enable` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for interlocked additions.

just-harry committed:
Implement the MSVC intrinsics for interlocked and-ing.

just-harry committed:
Implement the MSVC intrinsics for interlocked bit-test-and-resets.

just-harry committed:
Implement the MSVC intrinsics for interlocked bit-test-and-sets.

just-harry committed:
Implement the MSVC intrinsics for interlocked compare-and-exchanges.

just-harry committed:
Implement the MSVC intrinsics for 128-bit interlocked compare-and-exchanges.

just-harry committed:
Implement the MSVC intrinsics for interlocked pointer compare-and-exchanges.

just-harry committed:
Implement the MSVC intrinsics for interlocked decrements.

just-harry committed:
Implement the MSVC intrinsics for interlocked exchanges.

just-harry committed:
Implement the MSVC intrinsics for interlocked exchange-additions.

just-harry committed:
Implement the MSVC intrinsics for interlocked pointer exchanges.

just-harry committed:
Implement the MSVC intrinsics for interlocked increments.

just-harry committed:
Implement the MSVC intrinsics for interlocked or-ing.

just-harry committed:
Implement the MSVC intrinsics for interlocked xor-ing.

just-harry committed:
Implement the MSVC intrinsics for x86 I/O port handling.

just-harry committed:
Implement the `__int2c` MSVC intrinsic.

just-harry committed:
Implement the `__invlpg` MSVC intrinsic.

just-harry committed:
Implement the `__lidt` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for 64-bit bit-shifts.

just-harry committed:
Implement the MSVC intrinsics for `lzcnt`.

just-harry committed:
Implement the MSVC intrinsics for SSE float<->integer conversion.

just-harry committed:
Implement the MSVC intrinsics for some SSE4a instructions.

just-harry committed:
Implement the `_mm_stream_si64x` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for the x86 `rep movs` instruction.

just-harry committed:
Implement the `__noop` MSVC intrinsic.

just-harry committed:
Implement the `__nop` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for `popcnt`.

just-harry committed:
Implement the MSVC intrinsics for the x86 `rdtsc` instructions.

just-harry committed:
Implement the `__readcr<number>` MSVC intrinsics.

just-harry committed:
Implement the `__readdr` MSVC intrinsic.

just-harry committed:
Implement the `__readeflags` MSVC intrinsic.

just-harry committed:
Implement the `__readmsr` MSVC intrinsic.

just-harry committed:
Implement the `__readpmc` MSVC intrinsic.

just-harry committed:
Implement the `__segmentlimit` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for 128-bit bit-shifts.

just-harry committed:
Implement the `__sidt` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for the x86 `rep stos` instruction.

just-harry committed:
Implement the MSVC intrinsics for the AMD x86 SVM instruction set.

just-harry committed:
Implement the `__ud2` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for the Intel x86 VMX instruction set.

just-harry committed:
Implement the `__wbinvd` MSVC intrinsic.

just-harry committed:
Implement the `__writecr<number>` MSVC intrinsics.

just-harry committed:
Implement the `__writedr` MSVC intrinsic.

just-harry committed:
Implement the `__writeeflags` MSVC intrinsic.

just-harry committed:
Implement the `__writemsr` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for non-atomic read/write barriers.

just-harry committed:
Implement the MSVC intrinsics for bit-scanning.

just-harry committed:
Implement the MSVC intrinsics for bit-test operations.

just-harry committed:
Implement the MSVC intrinsics for byte-swapping.

just-harry committed:
Implement the MSVC intrinsics for bit-rotation.

just-harry committed:
Add a changelog entry for ImportC supporting some of MSVC's intrinsics.

just-harry committed:
Implement the `__assume` MSVC intrinsic.

just-harry commented on changelog/druntime.importc-initial-msvc-intrinsics.dd:
```diff
@@ -0,0 +1,4 @@
+ImportC, when targeting the Microsoft C runtime, supports a subset of the intrinsics recognised by the MSVC compiler.
```
Should American English spellings be used for changelog entries? ("recognised")

WalterBright replied:
yes

just-harry replied:
Noted. (That might be worth documenting in https://github.com/dlang/dmd/blob/75dca527f98b806fd82427d839f76df8290a1b00/changelog/README.md)

just-harry commented on druntime/src/__importc_msvc_builtins.d:
```diff
@@ -0,0 +1,12888 @@
+
+module __builtins_msvc;
+
+version (CRuntime_Microsoft)
+{
+    version = MSVCIntrinsics;
+}
```
Might `version (Windows)` be better for this? Or, is that too broad?

just-harry replied:
`CRuntime_Microsoft` is already used for the `__int<size>` types in `__importc_builtins.di`, so it probably suits here too.

thewilsonator commented on druntime/src/__importc_msvc_builtins.d:
```diff
@@ -0,0 +1,12888 @@
+                *remainder = cast(int) (dividend % divisor);
+                return cast(int) (dividend / divisor);
+            }
+            else
```
Is it really necessary to provide an asm implementation here? ditto throughout

just-harry replied:
For `_div64` specifically, yes, as compilers will generate the 64-bit encoding of `div`, which won't cause a #DE error on overflow of the 32-bit quotient, so a change of behaviour.

For everything else, generally yeah, either because it can't be represented without asm, or to ensure that the generated code behaves identically to MSVC's code, or for performance where it's relevant (like 128-bit multiplication).

thewilsonator commented on druntime/src/__importc_msvc_builtins.d:
```diff
@@ -0,0 +1,12888 @@
+        {
+            if (__ctfe)
+            {
+                /* This is an amalgamation of core.int128.divmod and core.int128.neg.  */
```
why not simply forward to that implementation?

just-harry replied:
I think that would rely on linking with DRuntime, which would rule out BetterC compatibility.

just-harry commented on compiler/src/dmd/cparse.d:
```diff
@@ -125,8 +125,8 @@ final class CParser(AST) : Parser!AST
-                    auto s = new AST.Import(Loc.initial, null, Id.importc_builtins, null, false);
-                    wrap.push(s);
+                    wrap.push(new AST.Import(Loc.initial, null, Id.importc_builtins, null, false));
+                    wrap.push(new AST.Import(Loc.initial, null, Id.importc_msvc_builtins, null, false));
```
Aye, they can be; good shout. (Though, the `__importc_msvc_builtins.d` file had to be renamed to `__builtins_msvc.d`, to match the module name.)

just-harry commented:
I've flipped this back to a draft PR as currently the functions in the `__builtins_msvc` module can be used only when the `-i` switch is supplied on the command-line.
I think it might be that semantic3 isn't getting run on that module's functions when `-i` isn't supplied—but I'm unsure.

WalterBright commented on changelog/druntime.importc-initial-msvc-intrinsics.dd:
```diff
@@ -0,0 +1,4 @@
+ImportC, when targeting the Microsoft C runtime, supports a subset of the intrinsics recognised by the MSVC compiler.
```
The documentation in the opening comment in the PR is far superior to this readme. Merge it in here.

just-harry replied:
I think I've copied the right part—better now?

WalterBright commented on druntime/src/importc_msvc_builtins_c.h:
```diff
@@ -0,0 +1,183 @@
+
+static inline unsigned char __readgsbyte(unsigned int Offset)
+{
+    return *(unsigned char __seg_gs *) (_import_c_msvc_ptr_int) Offset;
```
ImportC does not recognize __seg_gs or __seg_fs, so I don't see how this can work

just-harry replied:
I thought that GDC delegated to GCC for C files, but I was mistaken. I've replaced this with inline asm in D, as is already done for DMD.

WalterBright commented on compiler/src/dmd/cparse.d:
```diff
@@ -2728,8 +2728,8 @@ final class CParser(AST) : Parser!AST
-                if (idx.length > 2 && idx[0] == '_' && idx[1] == '_')  // leading double underscore
-                    importBuiltins = true;  // probably one of those compiler extensions
+                if (idx.length > 1 && idx[0] == '_')  // leading underscore
+                    importBuiltins = true;  // maybe one of those compiler extensions
```
I'm a little concerned about nearly always importing 10,000 lines of code for Windows compilations.

just-harry replied:
Mhm, it's not great. How about what I've pushed now?
We have a lazily-initialised `StringTable` of all the MSVC intrinsic names that we have implemented.
If we're targeting the Microsoft C runtime, and we come across an identifier with a leading underscore, then we check if that identifier is a known MSVC intrinsic, and if it is we import the MSVC builtins and we stop checking future identifiers.

The downside there is we have to remember to update that `StringTable` if more MSVC intrinsics are implemented, and we do pay the minor cost of hashing and looking up identifiers with leading underscores.

just-harry commented:
This is still a draft as I'm unable to get the compiler to actually generate code for the `__builtins_msvc` module without its path being supplied on the command-line.

---

Presently, neither the `__builtins` module nor the `__builtins_msvc` module make their way into the array of modules passed into `generateCodeAndWrite`, which is fine for `__builtins`, but not for `__builtins_msvc` because of how the standard Windows headers use the builtins, and because DMD doesn't inline some of the `asm`-using functions.

This is effectively how `winnt.h` declares `UnsignedMultiply128`:
```c
unsigned long long UnsignedMultiply128 (unsigned long long Multiplier, unsigned long long Multiplicand, unsigned long long *HighProduct);

#define UnsignedMultiply128 _umul128

unsigned long long UnsignedMultiply128 (unsigned long long Multiplier, unsigned long long Multiplicand, unsigned long long *HighProduct);
```
which will expand to `unsigned long long _umul128 (unsigned long long Multiplier, unsigned long long Multiplicand, unsigned long long *HighProduct);`, causing the linker to expect code for `_umul128`.

But even if the Windows headers didn't do that, we still need code to be generated for `_umul128`, because DMD won't inline `_umul128`'s call to `multiplyWithDoubleWidthProduct` and the linker will expect code for _it_.

(`_umul128` isn't the only function affected.)

---

Right now I'm stuck, because I've yet to figure out a good way to get the `__builtins_msvc` module into the modules array passed into `generateCodeAndWrite` after it's conditionally imported if any MSVC intrinsics are detected.

The best I've come up with thus far is very hacky, and would require the same changes to be made in LDC and GDC (unless I'm mistaken), which is: in `Module.parseModule`, if the module identifier is `__builtins_msvc`, set the global `includeImports` to true, and if it was previously false set `includeByDefault` to false, then add `__builtins_msvc` to the `compiledImports` array.


In short, my question is: is there an existing way to make it so the `__builtins_msvc` module winds up getting passed to `generateCodeAndWrite` after it gets imported by `CParser.parseModule`?
And, if not, is there an acceptably non-hacky way to thread it through the compiler without requiring changes to LDC and GDC?

Any suggestions would be appreciated.

just-harry committed:
Implement the `_mm_pause` MSVC intrinsic.

just-harry committed:
Implement the `__yield` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for ARM memory barriers.

just-harry committed:
Implement the MSVC intrinsics for `tzcnt`.

just-harry committed:
Add a test for checking that MSVC intrinsics are actually available from C.

WalterBright commented:
I suggest having the user do the #include line as the first line in his C code. Or have the user pass it in with the dmd switch that passes switches to the C preprocessor. Will that work?

just-harry committed:
Pass the resolved path of `<path>/importc.h/../include` to the C preprocessor as an include path.

just-harry commented:
Okay; a separate PR adding a mechanism that this PR can use to force codegen for the `__builtins_msvc` module: https://github.com/dlang/dmd/pull/16443

Via something like this in `cparse.d`:
```d
         if (token.value == TOK._import) // import declaration extension
         {
             auto a = parseImport();
             if (a && a.length)
/*+*/        {
/*+*/            auto imp = (*a)[0].isImport();
/*+*/            imp.forceCodegen = imp.id == Id.builtins_msvc;
                 symbols.append(a);
/*+*/        }
             return;
         }
```

ibuclaw commented on druntime/src/include/importc_msvc_builtins.h:
```diff
@@ -0,0 +1,31 @@
+
+#if defined(__clang__)
+#define __assume(expression) __builtin_assume(expression)
+#elif defined(__GNUC__)
+#define __assume(expression) do {if (!(expression)) {__builtin_unreachable();}} while (0)
+#else
```
In case this is meant to be public, FYI I think importC would discard this macro, whereas the other definitions (clang) are simple enough to be exposed as templates.

thewilsonator replied:
Also `__assume` is supposed to be an optimiser hint, not an assert, which is what this implementation does.

ibuclaw replied:
(unreachable is a compile-time barrier, not an assert)

I'll take your word for the clang and !clang!gnu paths.

thewilsonator replied:
> (unreachable is a compile-time barrier, not an assert)

I know that. `assert` is (often) defined as `do { if (!cond) { /* print stuff*/ ; __builtin_unreachable();}} while (0)`. This is almost identical (modulo the print stuff).

Assume should be `#define assume(x) x`, i.e. a no-op.

just-harry replied:
I'm two-years late in responding to this, but `__assume` should not be a no-op when mimicking the behaviour of MSVC.

[MSVC makes `__assume` assume two roles](https://learn.microsoft.com/cpp/intrinsics/assume#:~:text=The%20%5F%5Fassume%280%29%20statement%20is%20a%20special%20case%2E%20Use%20%5F%5Fassume%280%29%20to%20indicate%20a%20code%20path%20that%20can%27t%20be%20reached): one role is as an optimisation hint; the other role—the more important role—is to denote unreachable code when the expression is statically-known to be false or zero.

To demonstrate, here is a [Compiler Explorer session that shows `__assume(0)` preventing MSVC/clang/gcc from emitting a warning due to a lack of a return-value in a non-void function](https://godbolt.org/z/39d7oMTo7).

ibuclaw commented:
> Though, the issue of these builtins only working when the `-i` switch is used still remains.
> 
> Is there an existing way to force codegen for an imported module, when the `-i` switch isn't used?
> 

Did you try converting these functions to templates?

ibuclaw commented on druntime/src/__builtins_msvc.d:
```diff
@@ -0,0 +1,13488 @@
+            private void llvm_arm_hint(int) @safe pure nothrow @nogc;
+        }
+    }
+    else version (GNU)
```
Thanks for taking the time to consider both ldc and gdc here, it's really appreciated

 I was just wondering - GCC doesn't support MSVC. So are these GNU version blocks dead on Mingw targets?

just-harry replied:
> Thanks for taking the time to consider both ldc and gdc here, it's really appreciated

:)
> So are these GNU version blocks dead on Mingw targets?

I think so. Unless GDC defines `CRuntime_Microsoft` when targeting MinGW?
I don't really know the story as far as GDC's support of Windows goes. So, I just implemented the GDC versions to save some other poor sod the hassle of doing it later it on.

I don't think there's anything preventing this implementation from working when targeting MinGW or Cygwin, if that's wanted?

just-harry commented:
> > Though, the issue of these builtins only working when the `-i` switch is used still remains.
> > Is there an existing way to force codegen for an imported module, when the `-i` switch isn't used?
> 
> Did you try converting these functions to templates?

I did—but no dice, unfortunately, even with all the affected functions being templates, the simplest possible usage still fails to link:
```d
// windows_h_c.c:
#include <importc_msvc_builtins.h>
#include <windows.h>

// windows_h.d:
pragma(lib, "MinCore");
import windows_h_c;
```

```
windows_h.obj : error LNK2019: unresolved external symbol _InterlockedExchangeAdd referenced in function _InlineInterlockedAdd
windows_h.obj : error LNK2019: unresolved external symbol _InterlockedExchangeAdd64 referenced in function _InlineInterlockedAdd64
windows_h.obj : error LNK2019: unresolved external symbol _mul128 referenced in function MultiplyExtract128
windows_h.obj : error LNK2019: unresolved external symbol __shiftright128 referenced in function MultiplyExtract128
windows_h.obj : error LNK2019: unresolved external symbol _umul128 referenced in function UnsignedMultiplyExtract128
windows_h.obj : error LNK2019: unresolved external symbol _ReadWriteBarrier referenced in function BarrierAfterRead
windows_h.obj : error LNK2019: unresolved external symbol __stosb referenced in function RtlSecureZeroMemory
windows_h.obj : error LNK2019: unresolved external symbol __readgsqword referenced in function NtCurrentTeb
windows_h.exe : fatal error LNK1120: 8 unresolved externals
```

WalterBright commented:
Just have the user add the #include line.

just-harry commented:
> Just have the user add the #include line.

Perhaps I'm misunderstanding you, but is that not what the current implementation is doing?
In the example here: https://github.com/dlang/dmd/pull/16372#issuecomment-2094644837

There, the user `#include`s `<importc_msvc_builtins.h>` before `<windows.h>`, which works perfectly well.
`importc_msvc_builtins.h` then does an `__import __builtins_msvc;`
The import of `__builtins_msvc` succeeds because `druntime/import` is in the compiler's import path.
But because the path to `__builtins_msvc.d` wasn't explicitly provided on the command-line, codegen doesn't take place for the `__builtins_msvc` module.
Which ultimately causes the linker to fail because the way the standard Windows headers uses some of these intrinsics causes the linker to expect a definition for them.

ibuclaw commented:
> windows_h.obj : error LNK2019: unresolved external symbol _InterlockedExchangeAdd referenced in function _InlineInterlockedAdd

I can try to synthesize this later (Linux). But it would be useful to know how these symbols are being referenced in the windows headers.

My immediate suspicion is either the address is taken, or windows.h adds its own extern declaration outside of importc msvc builtins.

just-harry commented:
> > windows_h.obj : error LNK2019: unresolved external symbol _InterlockedExchangeAdd referenced in function _InlineInterlockedAdd
> 
> I can try to synthesize this later (Linux). But it would be useful to know how these symbols are being referenced in the windows headers.
> 
> My immediate suspicion is either the address is taken, or windows.h adds its own extern declaration outside of importc msvc builtins.


I think it's the latter.
For example, for `__shiftright128` in `winnt.h`, there's
```c
#define ShiftRight128 __shiftright128
DWORD64
ShiftRight128 (
    _In_ DWORD64 LowPart,
    _In_ DWORD64 HighPart,
    _In_ BYTE  Shift
    );
#pragma intrinsic(__shiftright128)
```

Similarly, for `_mul128` in `winnt.h`:
```c
#define Multiply128 _mul128
LONG64
Multiply128 (
    _In_ LONG64 Multiplier,
    _In_ LONG64 Multiplicand,
    _Out_ LONG64 *HighProduct
    );
#pragma intrinsic(_mul128)
```

The same pattern is used for the rest of affected functions, in `winnt.h`; with the minor difference of no `#define` being used for `_ReadWriteBarrier`, `__stosb`, or `__readgsqword`.

WalterBright commented:
> But because the path to __builtins_msvc.d wasn't explicitly provided on the command-line, codegen doesn't take place for the __builtins_msvc module.

For now I suggest just asking the user to put it on the command line.

just-harry commented:
> > But because the path to __builtins_msvc.d wasn't explicitly provided on the command-line, codegen doesn't take place for the __builtins_msvc module.
> 
> For now I suggest just asking the user to put it on the command line.

If it comes down to it, yeah okay.
But I don't think it makes for a great user-experience, and the point of ImportC is to reduce friction, so I'd like to try one last thing.

With an addition to ImportC, we can make it so these intrinsics can be inlined, and so that they can be templates (removing the need for unconditional codegen for the `__builtins_msvc` module)—improving runtime performance, reducing compile-times, and making them easier to use: https://github.com/dlang/dmd/pull/16464

just-harry committed:
Implement the MSVC intrinsics for volatile stores and loads.

just-harry committed:
Implement the MSVC intrinsics for prefetching.

just-harry committed:
Implement the `_mm_clflush` MSVC intrinsic.

just-harry committed:
Implement the MSVC intrinsics for loading/storing the x86 MXCSR register.

just-harry committed:
Implement the MSVC intrinsics for x86 memory fences.

just-harry committed:
Implement the `<op>x18<size>` MSVC intrinsics.

just-harry committed:
Implement the MSVC intrinsics for AArch64 acquire/release atomic loads/stores.

just-harry commented:
<details>
<summary>While revisiting this PR in preparation for <a href="https://github.com/dlang/dmd/pull/16464">https://github.com/dlang/dmd/pull/16464</a> I found a few more intrinsics defined via MSVC's `#pragma intrinsic` directive, and so I implemented the ones that I could; those are listed in this expando:</summary>

- `__addx18byte`
- `__addx18dword`
- `__addx18qword`
- `__addx18word`
- `__incx18byte`
- `__incx18dword`
- `__incx18qword`
- `__incx18word`
- `__readx18byte`
- `__readx18dword`
- `__readx18qword`
- `__readx18word`
- `__writex18byte`
- `__writex18dword`
- `__writex18qword`
- `__writex18word`
- `__iso_volatile_load16`
- `__iso_volatile_load32`
- `__iso_volatile_load64`
- `__iso_volatile_load8`
- `__iso_volatile_store16`
- `__iso_volatile_store32`
- `__iso_volatile_store64`
- `__iso_volatile_store8`
- `__load_acquire16`
- `__load_acquire32`
- `__load_acquire64`
- `__load_acquire8`
- `__ldar16`
- `__ldar32`
- `__ldar64`
- `__ldar8`
- `__stlr16`
- `__stlr32`
- `__stlr64`
- `__stlr8`
- `__prefetch`
- `__prefetchw`
- `_m_prefetchw`
- `_mm_prefetch`
- `_mm_clflush`
- `_mm_getcsr`
- `_mm_setcsr`
- `_mm_lfence`
- `_mm_mfence`
- `_mm_sfence`
</details>

There are a number of AArch64/ARM32 intrinsics which remain unimplemented for one of two reasons. \
The first being that they require compiler support, those being: [AArch64: `__getCallerReg`; `__getCallerRegFp`; `__setCallerReg`; `__setCallerRegFp`]. \
The second being that ImportC is unable to supply compile-time arguments to `__import`ed D function templates: [AArch64: `__break`; `__prefetch2`; `__getReg`; `__getRegFp`; `__setReg`; `__setRegFp`]; [ARM32: `__emit`; `_MoveFromCoprocessor`; `_MoveToCoprocessor`]; [AArch64/ARM32: `_ReadStatusReg`; `_WriteStatusReg`].

(As an aside, it would be really nice if ImportC had something like `__call_d(functionName)(templateArgs...)(runtimeArgs...)` which would call an `__import`ed D function with the usual IFTI rules, and something like `__instantiate(templateName)(templateArgs...)` for types.)

Additionally, I stripped out all usage of the `shared` type-qualifier for the various interlocked/atomic MSVC intrinsics because it didn't really work for C callers.

A small handful of unittests, which rely on DRuntime, have been disabled when compiling under `D_BetterC`.

Lastly, the changelog entry has been amended so that it shows how to use MSVC intrinsics in ImportC, and to point out that they enable `#include <windows.h>` to just work™.

just-harry commented:
To summarise the PR body and the last thirty-odd comments:

This PR implements a ton of the MSVC compiler's intrinsic functions. \
This is desirable because it greatly increases the likelihood of ImportC being able to successfully import and use C libraries that target Windows, especially ones focussed on achieving high performance (i.e. the ones that are worth using :)

Notably, this PR makes it possible for ImportC to successfully include the headers of the Windows SDK—which is unspeakably useful for Windows application and driver development. \
This addresses a common stumbling-block with ImportC: [1](https://github.com/dlang/dmd/issues/21815), [2](https://forum.dlang.org/post/osdpggewnyqlrzsiqlap@forum.dlang.org), [3](https://forum.dlang.org/post/wkqvhplpdibucepdtfle@forum.dlang.org), [4](https://forum.dlang.org/post/swmmlsaoefhrtansanfy@forum.dlang.org), [5](https://bugzilla-archive.dlang.org/bugs/23894/).

MSVC intrinsics can be used in ImportC by including the `importc_msvc_builtins.h` header.
The reason why these intrinsics aren't available automatically (like the intrinsics in DRuntime's `__builtins.di` file) is [to avoid the cost of unconditionally importing over 10,000 lines of code for most every Windows compilation (when using ImportC)](https://github.com/dlang/dmd/pull/16372#pullrequestreview-2002639550).

However, the Windows SDK's headers implement MSVC's intrinsics by first declaring them as external body-less functions, and then applying the `#pragma intrinsic` directive to undo that.
As ImportC does not understand `#pragma intrinsic`, when the Windows SDK's headers are included the linker ends up expecting external definitions for the MSVC intrinsics: which means that codegen has to take place unconditionally for all 461 intrinsics in the `__builtins_msvc` module, and to make matters worse, the intrinsics can't be inlined.

As such, this PR is currently blocked. (Codegen cannot take place unconditionally unless the `-i` switch is passed to the compiler, or if the path of `__builtins_msvc.d` is supplied to the compiler, both of which are unreasonable requirements.)

If https://github.com/dlang/dmd/pull/16464 is merged, `#pragma function_decl(ignore, ...)` can be used in `importc_msvc_builtins.h` to the same effect as `#pragma intrinsic`, thus side-stepping our codegen woes and allowing MSVC intrinsics to be inlined.

<details>
<summary>The full list of implemented intrinsics can be found in this expando:</summary>

- `__assume`
- `__iso_volatile_load8`
- `__iso_volatile_load16`
- `__iso_volatile_load32`
- `__iso_volatile_load64`
- `__iso_volatile_store8`
- `__iso_volatile_store16`
- `__iso_volatile_store32`
- `__iso_volatile_store64`
- `__debugbreak`
- `__fastfail`
- `_disable`
- `_enable`
- `_interlockedadd`
- `_interlockedadd64`
- `_InterlockedAnd`
- `_InterlockedAnd8`
- `_InterlockedAnd16`
- `_interlockedand64`
- `_interlockedbittestandreset`
- `_interlockedbittestandset`
- `_InterlockedCompareExchange`
- `_InterlockedCompareExchange8`
- `_InterlockedCompareExchange16`
- `_InterlockedCompareExchange64`
- `_InterlockedCompareExchangePointer`
- `_InterlockedDecrement`
- `_InterlockedDecrement16`
- `_interlockeddecrement64`
- `_InterlockedExchange`
- `_InterlockedExchange8`
- `_InterlockedExchange16`
- `_interlockedexchange64`
- `_InterlockedExchangeAdd`
- `_InterlockedExchangeAdd8`
- `_InterlockedExchangeAdd16`
- `_interlockedexchangeadd64`
- `_InterlockedExchangePointer`
- `_InterlockedIncrement`
- `_InterlockedIncrement16`
- `_interlockedincrement64`
- `_InterlockedOr`
- `_InterlockedOr8`
- `_InterlockedOr16`
- `_interlockedor64`
- `_InterlockedXor`
- `_InterlockedXor8`
- `_InterlockedXor16`
- `_interlockedxor64`
- `__noop`
- `__nop`
- `_ReadBarrier`
- `_WriteBarrier`
- `_ReadWriteBarrier`
- `_BitScanForward`
- `_BitScanReverse`
- `_bittest`
- `_bittestandcomplement`
- `_bittestandreset`
- `_bittestandset`
- `_byteswap_uint64`
- `_byteswap_ulong`
- `_byteswap_ushort`
- `_lrotr`
- `_lrotl`
- `_rotr`
- `_rotl`
- `_rotr64`
- `_rotl64`
- `_rotr16`
- `_rotl16`
- `_rotr8`
- `_rotl8`
- `_InterlockedAnd64`
- `_InterlockedDecrement64`
- `_InterlockedExchange64`
- `_InterlockedExchangeAdd64`
- `_InterlockedIncrement64`
- `_InterlockedOr64`
- `_InterlockedXor64`
- `__umulh`
- `__mulh`
- `_interlockedbittestandreset64`
- `_interlockedbittestandset64`
- `_InterlockedCompareExchange128`
- `_BitScanForward64`
- `_BitScanReverse64`
- `_bittest64`
- `_bittestandcomplement64`
- `_bittestandreset64`
- `_bittestandset64`
- `__emul`
- `__emulu`
- `_div64`
- `_udiv64`
- `_mm_pause`
- `_m_prefetchw`
- `_mm_prefetch`
- `_mm_clflush`
- `_mm_getcsr`
- `_mm_setcsr`
- `_mm_lfence`
- `_mm_mfence`
- `_mm_sfence`
- `__cpuid`
- `__cpuidex`
- `_cvt_ftoi_fast`
- `_cvt_ftoll_fast`
- `_cvt_ftoui_fast`
- `_cvt_ftoull_fast`
- `_cvt_dtoi_fast`
- `_cvt_dtoll_fast`
- `_cvt_dtoui_fast`
- `_cvt_dtoull_fast`
- `_cvt_ftoi_sat`
- `_cvt_ftoll_sat`
- `_cvt_ftoui_sat`
- `_cvt_ftoull_sat`
- `_cvt_dtoi_sat`
- `_cvt_dtoll_sat`
- `_cvt_dtoui_sat`
- `_cvt_dtoull_sat`
- `_cvt_ftoi_sent`
- `_cvt_ftoll_sent`
- `_cvt_ftoui_sent`
- `_cvt_ftoull_sent`
- `_cvt_dtoi_sent`
- `_cvt_dtoui_sent`
- `_cvt_dtoll_sent`
- `_cvt_dtoull_sent`
- `__halt`
- `_InterlockedAnd_HLEAcquire`
- `_InterlockedAnd_HLERelease`
- `_interlockedbittestandreset_HLEAcquire`
- `_interlockedbittestandreset_HLERelease`
- `_interlockedbittestandset_HLEAcquire`
- `_interlockedbittestandset_HLERelease`
- `_InterlockedCompareExchange_HLEAcquire`
- `_InterlockedCompareExchange_HLERelease`
- `_InterlockedCompareExchange64_HLEAcquire`
- `_InterlockedCompareExchange64_HLERelease`
- `_InterlockedCompareExchangePointer_HLEAcquire`
- `_InterlockedCompareExchangePointer_HLERelease`
- `_InterlockedExchange_HLEAcquire`
- `_InterlockedExchange_HLERelease`
- `_InterlockedExchangeAdd_HLEAcquire`
- `_InterlockedExchangeAdd_HLERelease`
- `_InterlockedExchangePointer_HLEAcquire`
- `_InterlockedExchangePointer_HLERelease`
- `_InterlockedOr_HLEAcquire`
- `_InterlockedOr_HLERelease`
- `_InterlockedXor_HLEAcquire`
- `_InterlockedXor_HLERelease`
- `__inbyte`
- `__inword`
- `__indword`
- `__outbyte`
- `__outword`
- `__outdword`
- `__inbytestring`
- `__inwordstring`
- `__indwordstring`
- `__outbytestring`
- `__outwordstring`
- `__outdwordstring`
- `__int2c`
- `__invlpg`
- `__lidt`
- `__ll_lshift`
- `__ll_rshift`
- `__ull_rshift`
- `__lzcnt16`
- `__lzcnt`
- `_lzcnt_u32`
- `_tzcnt_u16`
- `_tzcnt_u32`
- `_mm_extract_si64`
- `_mm_extracti_si64`
- `_mm_insert_si64`
- `_mm_inserti_si64`
- `_mm_stream_sd`
- `_mm_stream_ss`
- `__movsb`
- `__movsw`
- `__movsd`
- `__popcnt16`
- `__popcnt`
- `__rdtsc`
- `__rdtscp`
- `__readcr0`
- `__readcr2`
- `__readcr3`
- `__readcr4`
- `__readcr8`
- `__readdr`
- `__readeflags`
- `__readmsr`
- `__readpmc`
- `__segmentlimit`
- `__sidt`
- `__stosb`
- `__stosw`
- `__stosd`
- `__svm_clgi`
- `__svm_invlpga`
- `__svm_skinit`
- `__svm_stgi`
- `__svm_vmload`
- `__svm_vmrun`
- `__svm_vmsave`
- `__ud2`
- `__vmx_off`
- `__vmx_vmptrst`
- `__wbinvd`
- `__writecr0`
- `__writecr2`
- `__writecr3`
- `__writecr4`
- `__writecr8`
- `__writedr`
- `__writeeflags`
- `__writemsr`
- `_umul128`
- `_mul128`
- `_div128`
- `_udiv128`
- `__readgsbyte`
- `__readgsword`
- `__readgsdword`
- `__readgsqword`
- `__writegsbyte`
- `__writegsword`
- `__writegsdword`
- `__writegsqword`
- `__addgsbyte`
- `__addgsword`
- `__addgsdword`
- `__addgsqword`
- `__incgsbyte`
- `__incgsword`
- `__incgsdword`
- `__incgsqword`
- `__faststorefence`
- `_InterlockedAnd_np`
- `_InterlockedAnd8_np`
- `_InterlockedAnd16_np`
- `_InterlockedAnd64_np`
- `_InterlockedAnd64_HLEAcquire`
- `_InterlockedAnd64_HLERelease`
- `_interlockedbittestandreset64_HLEAcquire`
- `_interlockedbittestandreset64_HLERelease`
- `_interlockedbittestandset64_HLEAcquire`
- `_interlockedbittestandset64_HLERelease`
- `_InterlockedCompareExchange_np`
- `_InterlockedCompareExchange16_np`
- `_InterlockedCompareExchange64_np`
- `_InterlockedCompareExchange128_np`
- `_InterlockedCompareExchangePointer_np`
- `_InterlockedExchange64_HLEAcquire`
- `_InterlockedExchange64_HLERelease`
- `_InterlockedExchangeAdd64_HLEAcquire`
- `_InterlockedExchangeAdd64_HLERelease`
- `_InterlockedOr_np`
- `_InterlockedOr8_np`
- `_InterlockedOr16_np`
- `_InterlockedOr64_np`
- `_InterlockedOr64_HLEAcquire`
- `_InterlockedOr64_HLERelease`
- `_InterlockedXor_np`
- `_InterlockedXor8_np`
- `_InterlockedXor16_np`
- `_InterlockedXor64_np`
- `_InterlockedXor64_HLEAcquire`
- `_InterlockedXor64_HLERelease`
- `__lzcnt64`
- `_lzcnt_u64`
- `_tzcnt_u64`
- `_mm_cvtsi64x_ss`
- `_mm_cvtss_si64x`
- `_mm_cvttss_si64x`
- `_mm_stream_si64x`
- `__movsq`
- `__popcnt64`
- `__shiftleft128`
- `__shiftright128`
- `__stosq`
- `__vmx_on`
- `__vmx_vmclear`
- `__vmx_vmlaunch`
- `__vmx_vmptrld`
- `__vmx_vmread`
- `__vmx_vmresume`
- `__vmx_vmwrite`
- `__readfsbyte`
- `__readfsword`
- `__readfsdword`
- `__readfsqword`
- `__writefsbyte`
- `__writefsword`
- `__writefsdword`
- `__writefsqword`
- `__addfsbyte`
- `__addfsword`
- `__addfsdword`
- `__incfsbyte`
- `__incfsword`
- `__incfsdword`
- `_InterlockedAddLargeStatistic`
- `__prefetch`
- `__dmb`
- `__dsb`
- `__isb`
- `_InterlockedAdd`
- `_InterlockedAdd_acq`
- `_InterlockedAdd_rel`
- `_InterlockedAdd_nf`
- `_InterlockedAdd64`
- `_InterlockedAdd64_acq`
- `_InterlockedAdd64_rel`
- `_InterlockedAdd64_nf`
- `_InterlockedAnd_acq`
- `_InterlockedAnd_rel`
- `_InterlockedAnd_nf`
- `_InterlockedAnd8_acq`
- `_InterlockedAnd8_rel`
- `_InterlockedAnd8_nf`
- `_InterlockedAnd16_acq`
- `_InterlockedAnd16_rel`
- `_InterlockedAnd16_nf`
- `_InterlockedAnd64_acq`
- `_InterlockedAnd64_rel`
- `_InterlockedAnd64_nf`
- `_interlockedbittestandreset_acq`
- `_interlockedbittestandreset_rel`
- `_interlockedbittestandreset_nf`
- `_interlockedbittestandset_acq`
- `_interlockedbittestandset_rel`
- `_interlockedbittestandset_nf`
- `_InterlockedCompareExchange_acq`
- `_InterlockedCompareExchange_rel`
- `_InterlockedCompareExchange_nf`
- `_InterlockedCompareExchange8_acq`
- `_InterlockedCompareExchange8_rel`
- `_InterlockedCompareExchange8_nf`
- `_InterlockedCompareExchange16_acq`
- `_InterlockedCompareExchange16_rel`
- `_InterlockedCompareExchange16_nf`
- `_InterlockedCompareExchange64_acq`
- `_InterlockedCompareExchange64_rel`
- `_InterlockedCompareExchange64_nf`
- `_InterlockedCompareExchangePointer_acq`
- `_InterlockedCompareExchangePointer_rel`
- `_InterlockedCompareExchangePointer_nf`
- `_InterlockedDecrement_acq`
- `_InterlockedDecrement_rel`
- `_InterlockedDecrement_nf`
- `_InterlockedDecrement16_acq`
- `_InterlockedDecrement16_rel`
- `_InterlockedDecrement16_nf`
- `_InterlockedDecrement64_acq`
- `_InterlockedDecrement64_rel`
- `_InterlockedDecrement64_nf`
- `_InterlockedExchange_acq`
- `_InterlockedExchange_rel`
- `_InterlockedExchange_nf`
- `_InterlockedExchange8_acq`
- `_InterlockedExchange8_rel`
- `_InterlockedExchange8_nf`
- `_InterlockedExchange16_acq`
- `_InterlockedExchange16_rel`
- `_InterlockedExchange16_nf`
- `_InterlockedExchange64_acq`
- `_InterlockedExchange64_rel`
- `_InterlockedExchange64_nf`
- `_InterlockedExchangeAdd_acq`
- `_InterlockedExchangeAdd_rel`
- `_InterlockedExchangeAdd_nf`
- `_InterlockedExchangeAdd8_acq`
- `_InterlockedExchangeAdd8_rel`
- `_InterlockedExchangeAdd8_nf`
- `_InterlockedExchangeAdd16_acq`
- `_InterlockedExchangeAdd16_rel`
- `_InterlockedExchangeAdd16_nf`
- `_InterlockedExchangeAdd64_acq`
- `_InterlockedExchangeAdd64_rel`
- `_InterlockedExchangeAdd64_nf`
- `_InterlockedExchangePointer_acq`
- `_InterlockedExchangePointer_rel`
- `_InterlockedExchangePointer_nf`
- `_InterlockedIncrement_acq`
- `_InterlockedIncrement_rel`
- `_InterlockedIncrement_nf`
- `_InterlockedIncrement16_acq`
- `_InterlockedIncrement16_rel`
- `_InterlockedIncrement16_nf`
- `_InterlockedIncrement64_acq`
- `_InterlockedIncrement64_rel`
- `_InterlockedIncrement64_nf`
- `_InterlockedOr_acq`
- `_InterlockedOr_rel`
- `_InterlockedOr_nf`
- `_InterlockedOr8_acq`
- `_InterlockedOr8_rel`
- `_InterlockedOr8_nf`
- `_InterlockedOr16_acq`
- `_InterlockedOr16_rel`
- `_InterlockedOr16_nf`
- `_InterlockedOr64_acq`
- `_InterlockedOr64_rel`
- `_InterlockedOr64_nf`
- `_InterlockedXor_acq`
- `_InterlockedXor_rel`
- `_InterlockedXor_nf`
- `_InterlockedXor8_acq`
- `_InterlockedXor8_rel`
- `_InterlockedXor8_nf`
- `_InterlockedXor16_acq`
- `_InterlockedXor16_rel`
- `_InterlockedXor16_nf`
- `_InterlockedXor64_acq`
- `_InterlockedXor64_rel`
- `_InterlockedXor64_nf`
- `__yield`
- `__readx18byte`
- `__readx18word`
- `__readx18dword`
- `__readx18qword`
- `__writex18byte`
- `__writex18word`
- `__writex18dword`
- `__writex18qword`
- `__addx18byte`
- `__addx18word`
- `__addx18dword`
- `__addx18qword`
- `__incx18byte`
- `__incx18word`
- `__incx18dword`
- `__incx18qword`
- `_interlockedbittestandreset64_acq`
- `_interlockedbittestandreset64_rel`
- `_interlockedbittestandreset64_nf`
- `_interlockedbittestandset64_acq`
- `_interlockedbittestandset64_rel`
- `_interlockedbittestandset64_nf`
- `_InterlockedCompareExchange128_acq`
- `_InterlockedCompareExchange128_rel`
- `_InterlockedCompareExchange128_nf`
- `__ldar8`
- `__ldar16`
- `__ldar32`
- `__ldar64`
- `__load_acquire8`
- `__load_acquire16`
- `__load_acquire32`
- `__load_acquire64`
- `__stlr8`
- `__stlr16`
- `__stlr32`
- `__stlr64`
- `__prefetchw`
</details>

just-harry commented:
And one last thing.
I understand that by contributing these changes to DMD I assign the copyright for the code to the D Language Foundation, who then license the code under the Boost Software Licence.

I would like to request that the code I authored, in the `druntime/src/__builtins_msvc.d` file, be additionally/alternatively licensed under the [Zero Clause BSD Licence](https://choosealicense.com/licenses/0bsd/), please. \
That is to say, that code would be available for use under _either_ the Boost Software Licence OR the Zero Clause BSD Licence.

The Boost Software Licence has only one condition, which is that copies of the source code must retain the copyright notice. \
The Zero Clause BSD Licence has no conditions, so it imposes no additional restrictions.

Would this be acceptable, @WalterBright? \
Thank you, in advance, for your consideration.

WalterBright commented:
> Would this be acceptable, @WalterBright?

yes, it is acceptable.

just-harry committed:
Document the ImportC `include` directory in ImportC's specification.

just-harry committed:
Document the `__builtins_msvc` module in ImportC's specification.

just-harry committed:
Apply `#pragma function_decl(ignore)` to the MSVC intrinsics implemented in `__builtins_msvc`.

just-harry committed:
Document the alternative licence of 0BSD for the MSVC intrinsics implemented in `__builtins_msvc.d`.

See https://github.com/dlang/dmd/pull/16372#issuecomment-4819698575
and https://github.com/dlang/dmd/pull/16372#issuecomment-4858357909

just-harry committed:
Define `__IMPORTC_DMD__` in `importc.h` to make it possible to detect DMD via the C preprocessor.

just-harry commented:
> > Would this be acceptable, @WalterBright?
>
> yes, it is acceptable.

Wonderful! Thank you; it's greatly appreciated.

I've documented this in a comment near to the top of the `__builtins_msvc.d` file.

just-harry commented:
Now that https://github.com/dlang/dmd/pull/16464 has been merged, I've amended `importc_msvc_builtins.h` to apply `#pragma function_decl(ignore)` to all the implemented MSVC intrinsics: this allows the intrinsics to be inlined, and solves our codegen woes.

Additionally, the new `include` directory is now documented in ImportC's specification, and the same goes for the `__builtins_msvc.d` file/module and the `importc_msvc_builtins.h` header.

---

There is one final snag preventing this PR from leaving draft status:
MSVC's `__assume` intrinsic is both an optimisation-hint and MSVC's mean of denoting unreachable code;
as such it has a semantic effect and should not be a no-op.

As it denotes unreachable code, it is my understanding that it must be a macro so that the supplied expression is available to the host compiler.

This macro's implementation depends on the host D compiler: but there is no way to determine the host D compiler via the C preprocessor.

I propose a simple solution:
- DMD's `importc.h` file should define `__IMPORTC_DMD__`.
- LDC's `importc.h` file should define `__IMPORTC_LDC__`.
- GDC does not have an `importc.h` file (as far as I can tell), nor does it define `__IMPORTC__`.

Thus, DMD can be be detected via `__IMPORTC_DMD__`, LDC via `__IMPORTC_LDC__`, and GDC via the absence of `__IMPORTC__`.

I've implemented `__IMPORTC_DMD__` [in this PR](https://github.com/dlang/dmd/pull/16372/changes/dbe5690eefaac8275d0237a47cda787936616d91), and submitted a [PR for `__IMPORTC_LDC__` here](https://github.com/ldc-developers/ldc/pull/5167).

@WalterBright, @kinke, any objections to defining `__IMPORTC_DMD__` and `__IMPORTC_LDC__`?

thewilsonator reviewed (commented):
```
/tmpfs/dautotest-work/repo/dlang.org/web/spec/importc.html:145: [RAW MACRO LEAKAGE] $(SPEC_S ImportC,
/tmpfs/dautotest-work/repo/dlang.org/web/dlangspec.html:40232: [RAW MACRO LEAKAGE] $(SPEC_S ImportC,
2 errors found. Exiting.
```
is coming somewhere from the spec changes but I can't see where.

kinke commented on compiler/src/dmd/link.d:
```diff
@@ -1026,6 +1019,14 @@ public int runPreprocessor(Loc loc, const(char)[] cpp, const(char)[] filename, c
+                if (includePath)
+                {
+                    buf.write(" /I");
+                    buf.write(includePath);
```
You might need to quote here.

just-harry replied:
Indeed I do; thanks for catching that!

kinke commented on druntime/src/__builtins_msvc.d:
AFAICT, this module isn't (pre)compiled and linked, so should probably have a `.di` extension.

just-harry replied:
Given that all the public symbols are now templates, that makes sense—but doing so causes the file's unittests to have no effect at runtime... so I'm going to not do that unless given a compelling reason :)

just-harry commented:
> ```
> /tmpfs/dautotest-work/repo/dlang.org/web/spec/importc.html:145: [RAW MACRO LEAKAGE] $(SPEC_S ImportC,
> /tmpfs/dautotest-work/repo/dlang.org/web/dlangspec.html:40232: [RAW MACRO LEAKAGE] $(SPEC_S ImportC,
> 2 errors found. Exiting.
> ```
> 
> is coming somewhere from the spec changes but I can't see where.

A paragraph was missing its closing bracket: fixed now

just-harry commented:
LDC [now defines `__IMPORTC_LDC__`](https://github.com/ldc-developers/ldc/pull/5167), and I'm going to take it that there are no qualms to defining `__IMPORTC_DMD__`, so `__assume` is grand now.

I've double-checked that all the intrinsics work as they should when compiled with DMD, LDC, and GDC.

---

I now consider this PR fully ready for final review (and hopefully merging).

As a reminder, I've split the implementation of the intrinsics up into a few-dozen commits to aid in reviewing them.

(P.S. please don't squash-merge this, no one wants a single commit that big!)

just-harry commented on compiler/src/dmd/link.d:
```diff
@@ -1026,6 +1027,13 @@ public int runPreprocessor(Loc loc, const(char)[] cpp, const(char)[] filename, c
                 buf.printf(" /std:c11 /P /Zc:preprocessor /PD /nologo /utf-8 \"%.*s\" \"/FI%s\" \"/Fi%.*s\"",
                     cast(int)filename.length, filename.ptr, importc_h, cast(int)output.length, output.ptr);
 
+                /* Append the include path, if it was provided
+                 */
+                if (includePath)
+                {
+                    buf.printf(` "/I%.*s"`, cast(int)includePath.length, includePath.ptr);
+                }
```
This change will need to be made in LDC also.

I assume the same applies to GDC, although as I understand it GDC doesn't support targeting Windows, so whether GDC implements this change or not is immaterial at present.

To clarify, the reason for this change is to make `importc_msvc_builtins.h` available to the C preprocessor.

In future, other ImportC/DRuntime-provided headers can be placed in this directory, so this is a one-time change.

Pre-emptive LDC PR: https://github.com/ldc-developers/ldc/pull/5169

thewilsonator commented:
> (P.S. please don't squash-merge this, no one wants a single commit that big!)

I'd rather not have 89 commits though.

just-harry commented:
> > (P.S. please don't squash-merge this, no one wants a single commit that big!)
> 
> I'd rather not have 89 commits though.

That's understandable.  I (or someone else) can rebase the commits into two-or-three commits after review?

I think the massive `__builtins_msvc.d` commit should be separate from the compiler changes, at least.

thewilsonator commented:
Sounds good.

dkorpel commented:
The new file is only added to druntime/mak/COPY and not druntime/mak/SRCS, so I think the unittests aren't run by CI currently.

just-harry committed:
Ensure that unittests are run for `__builtins_msvc.d`.

This commit will later be squashed into "Add files for MSVC intrinsic implementations.".

just-harry committed:
Remove an unused import of `std.math.nextUp`.

just-harry committed:
Handle `prefetch` more reliably.

just-harry committed:
Fix a unittest failing to compile without `-dip1000` when targeting x86.

just-harry commented:
> The new file is only added to druntime/mak/COPY and not druntime/mak/SRCS, so I think the unittests aren't run by CI currently.

@dkorpel, thanks for spotting that. You're correct about that causing the CI to not run the unittests. \
It's fixed now.

I was following the lead of `__importc_builtins.di` which also isn't listed in `SRCS`.

rainers commented:
Building druntime unittests with `-wi` shows that dmd is unable to inline a lot of these functions (about 25 templates affected), e.g.
```
src\__builtins_msvc.d(1635): Warning: cannot inline function `__builtins_msvc._mm_getcsr!()`
        uint _mm_getcsr()() @trusted nothrow @nogc
             ^
src\__builtins_msvc.d(1676): Warning: cannot inline function `__builtins_msvc._mm_setcsr!()`
        void _mm_setcsr()(uint MxCsr) @trusted nothrow @nogc
             ^
src\__builtins_msvc.d(2016): Warning: cannot inline function `__builtins_msvc.cpuID!()`
        private void cpuID(Args...)(scope int[4]* cpuInfo, int function_id, Args args) @safe pure nothrow @nogc
                     ^
src\__builtins_msvc.d(4847): Warning: cannot inline function `__builtins_msvc.manipulateMemoryThroughTIBSegmentRegister!(uint, null)`
        manipulateMemoryThroughTIBSegmentRegister(
        ^
```
Maybe better omit the `pragma(inline, true)` in these cases and let the compiler decide silently where it inlines functions? Sometimes inlining can be achieved by changing the code a bit, e.g. converting `switch` to a list of ternary expressions.

just-harry commented:
@rainers,
For LDC and GDC, I'd personally prefer to ensure that these functions are always inlined. \
But I recognise that the wall of warnings when using DMD is annoying, so how about [a compromise: `pragma(inline, true)` for not DMD](https://github.com/dlang/dmd/pull/23616)?
